# Content


* 1. Import Liraries

* 2. Set working path

* 3. Import data

* 4. Merge data frames

* 5. Export to pkl format

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os
import dask.dataframe as dd

In [3]:
# setting working path

path= r'/Users/phongsathorn/Documents/Data Analytics/CareerFoundry/Achievement 4 - Python/InstaCart/Instacart Business Analysis'

In [4]:
# importing a data file of orders_checked.csv and orders_products_prior

df_ords_prior=pd.read_csv(os.path.join(path,'02 Data','Original Data','orders_products_prior.csv'))
df_ords=pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_checked.csv'))

In [5]:
## check for heads

df_ords.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             1           1   2398795        1             2   
2             2           2    473747        1             3   
3             3           3   2254736        1             4   
4             4           4    431534        1             5   

   orders_day_of_week  time_of_day  days_since_previous_order  
0                   2            8                        NaN  
1                   3            7                       15.0  
2                   3           12                       21.0  
3                   4            7                       29.0  
4                   4           15                       28.0

In [6]:
## Check for head

df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [7]:
# Check for number of rows and columns 

df_ords.shape

(3421083, 8)

In [8]:
df_ords_prior.shape

(32434489, 4)

Merging 2 dataframes together , using "order_id" as key-column, using "inner join".
There are 2 different commands possible : 
1. df_merge= df_ords.merge(df_ords_prior, on = ['order_id'], indicator=True)
2. merged_ddf = dd.merge(df_ords, df_ords_prior, how='inner', on='order_id', indicator=True)

Note: 
Using "inner join" will only merge/combine values that exist in both dataframes together.
If you want to check for a full match, you need to use the argument "how= 'outer'"

# 4. Merge data frames

In [9]:
# Using merge() command
df_merge= df_ords.merge(df_ords_prior, on = ['order_id'],how='inner', indicator=True)

In [10]:
df_merge.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             0           0   2539329        1             1   
2             0           0   2539329        1             1   
3             0           0   2539329        1             1   
4             0           0   2539329        1             1   

   orders_day_of_week  time_of_day  days_since_previous_order  product_id  \
0                   2            8                        NaN         196   
1                   2            8                        NaN       14084   
2                   2            8                        NaN       12427   
3                   2            8                        NaN       26088   
4                   2            8                        NaN       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [15]:
# drop columns

df_merge=df_merge.drop(columns=['Unnamed: 0.1','Unnamed: 0'])

In [18]:
# Using merge() command with how= ''
# merged_ddf = dd.merge(df_ords, df_ords_prior, how='inner', on='order_id', indicator=True)

In [18]:
df_merge.head()

order_id  user_id  order_number  orders_day_of_week  time_of_day  \
0   2539329        1             1                   2            8   
1   2539329        1             1                   2            8   
2   2539329        1             1                   2            8   
3   2539329        1             1                   2            8   
4   2539329        1             1                   2            8   

   days_since_previous_order  product_id  add_to_cart_order  reordered _merge  
0                        NaN         196                  1          0   both  
1                        NaN       14084                  2          0   both  
2                        NaN       12427                  3          0   both  
3                        NaN       26088                  4          0   both  
4                        NaN       26405                  5          0   both

In [20]:
# counting the number of 'both' in the ['_merge'] column
df_merge['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

Only "inner join" was used here, thats why we have 32434489 matches. This means: each of those rows have information found in both input data sets.

Using the argument of "how= 'outer'" will give another number. This number will also show rows that does not match both dataframes.

# 5. Export to pkl format

In [25]:
# Export data to pkl format

df_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))

In [ ]:
# when importing pkl format
pd.read_pkl()

In [26]:
df_merge.shape

(32434489, 10)